File: src/main/scala/psf/FloatPartCSRResult.java
package psf;

import com.tencent.angel.exception.AngelException;
import com.tencent.angel.ml.matrix.psf.get.base.PartitionGetResult;
import com.tencent.angel.ps.impl.matrix.ServerSparseFloatRow;
import com.tencent.angel.ps.impl.matrix.ServerRow;
import io.netty.buffer.ByteBuf;
import it.unimi.dsi.fastutil.ints.Int2FloatOpenHashMap;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import java.util.Arrays;
import java.util.List;
import java.util.Map;


/**
 * Created by chris on 9/19/17.
 */
public class FloatPartCSRResult extends PartitionGetResult{
    private static final Log LOG = LogFactory.getLog(FloatPartCSRResult.class);

    private List<ServerRow> splits;
    private ByteBuf buf;
    private int len;
    private int readerIdx;

    public FloatPartCSRResult(List<ServerRow> splits) {
        this.splits = splits;
    }

    public FloatPartCSRResult() {}

    @Override
    public void serialize(ByteBuf buf) {
        // Write #rows
        buf.writeInt(splits.size());
        // Write each row
        for (ServerRow row : splits) {
            if (row instanceof ServerSparseFloatRow) {
                serialize(buf, (ServerSparseFloatRow) row);
            } else {
                throw new AngelException("LDA should be set with ServerDenseIntRow");
            }
        }
    }

    public void serialize(ByteBuf buf, ServerSparseFloatRow row) {

        try {
            row.getLock().readLock().lock();
            Int2FloatOpenHashMap map = row.getData();
            int cnt = 0;

            for (Map.Entry<Integer, Float> entry : map.entrySet()) {
                if(entry.getValue() != 0.0) {
                    cnt ++;
                }
            }


            buf.writeInt(cnt);
            for (Map.Entry<Integer, Float> entry : map.entrySet()) {
                if(entry.getValue() != 0.0) {
                    buf.writeInt(entry.getKey());
                    buf.writeFloat(entry.getValue());
                }
            }
        } finally {
            row.getLock().readLock().unlock();
        }

    }

    @Override
    public void deserialize(ByteBuf buf) {
        this.len = buf.readInt();
        this.buf = buf.duplicate();
        this.buf.retain();
//    LOG.info(buf.refCnt());
        this.readerIdx = 0;
    }

    @Override
    public int bufferLen() {
        return 48;
    }

    public boolean read(float[] row) {
        if (readerIdx == len)
            return false;

        readerIdx ++;
                // sparse

        Arrays.fill(row, 0f);
        int len = buf.readInt();
        for (int i = 0; i < len; i ++) {
            int key = buf.readInt();
            float val = buf.readFloat();
            row[key] = val;
        }

        if (readerIdx == this.len) {
            buf.release();
//      LOG.info(buf.refCnt());
        }
        return true;
    }

}

File: src/main/scala/adni/psf/FloatPartCSRResult.scala
package adni.psf

import java.nio.FloatBuffer

import com.tencent.angel.ml.matrix.psf.get.base.PartitionGetResult
import com.tencent.angel.ps.impl.matrix.{ServerDenseFloatRow, ServerRow}
import io.netty.buffer.ByteBuf
import org.apache.commons.logging.LogFactory

import scala.collection.mutable

/**
  * Created by chris on 11/6/17.
  */
object FloatPartCSRResult {
  private val LOG = LogFactory.getLog(classOf[FloatPartCSRResult])
}

class FloatPartCSRResult() extends PartitionGetResult {
  private var row: ServerRow = _
  private var buf: ByteBuf = _
  private var len: Int = 0
  private var readerIdx: Int = 0

  def this(row: ServerRow) {
    this()
    this.row = row
  }

  override def serialize(buf: ByteBuf): Unit = { // Write #rows
    buf.writeInt(1)
    // Write each row
    serialize(buf, row.asInstanceOf[ServerDenseFloatRow])
  }

  def serialize(buf: ByteBuf, row: ServerDenseFloatRow): Unit = {
    try {
      row.getLock.readLock.lock()
      val m:FloatBuffer = row.getData
      val partLens:Int = (row.getEndCol - row.getStartCol).toInt
      var cnt = 0
      (0 until partLens) foreach{i =>
        if(m.get(i) > 0.0f)
          cnt += 1
      }
      buf.writeInt(cnt)
      (0 until partLens) foreach{i =>
          if (m.get(i) > 0.0f){
            buf.writeInt(row.getStartCol.toInt + i)
            buf.writeFloat(m.get(i))
          }
      }
    }
    finally {
      row.getLock.readLock.unlock()
    }
  }

  override def deserialize(buf: ByteBuf): Unit = {
    this.len = buf.readInt
    this.buf = buf.duplicate
    this.buf.retain
    //    LOG.info(buf.refCnt());
    this.readerIdx = 0
  }

  override def bufferLen: Int = {
     48
  }

  def read(row: mutable.Map[Int, Float]): Boolean = {
    if (readerIdx == this.len) {
      return false
    }
    readerIdx += 1
    // sparse
    row.clear()
    val len: Int = buf.readInt
    var i: Int = 0
    while (i < len) {
      val k: Int = buf.readInt
      val v: Float = buf.readFloat
      row.put(k, v)
      i += 1
    }
    if (readerIdx == this.len) {
      buf.release
      //      LOG.info(buf.refCnt());
    }
    true
  }
}
